### PR TITLE
[lte][cwf][orc8r] Split relay_enabled flag to separate support for hss and gx_gy

### PR DESCRIPTION
**Pull request by andreilee**
Signed-off-by: Andrei Lee <andreilee@fb.com>

## Summary

Second PR in a set of 2

Previous is https://github.com/magma/magma/pull/2575

Federation support is being split between gx_gy and hss. In terms of configuration, the `relay_enabled` flag is being split to support this, with new `gx_gy_relay_enabled` and `hss_relay_enabled` flags.

This second PR modifes the cloud-side for the relay split, and adds a migration script to run on the db.

- Split API endpoints for `relay_enabled` for `hss_relay_enabled` and `gx_gy_relay_enabled`
- Split support for `relay_enabled` in mconfigs
- Each gateway service is now changed for split relay support

### Verifying Migration Scripts

Migration script output from `m013_relay_split`:
```
root@a73c73ce798e:/var/opt/magma/bin# ./m013_relay_split
I0911 22:37:40.073581     338 main.go:68] BEGIN MIGRATION
I0911 22:37:40.073909     338 main.go:72] SQL_DRIVER: postgres
I0911 22:37:40.073949     338 main.go:73] DATABASE_SOURCE: dbname=pgdump-dev-9-4-20 user=magma_dev password=magma_dev host=postgres sslmode=disable
I0911 22:37:40.079857     338 main.go:114] [RUN] SELECT network_id, value FROM cfg_network_configs WHERE type = $1 [cellular_network]
I0911 22:37:40.081731     338 main.go:132] Old config: {Epc:0xc000372000 FegNetworkID:[] Ran:[123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]}
I0911 22:37:40.081873     338 main.go:133] Old config epc: &{CloudSubscriberdbEnabled:false DefaultRuleID:default_rule_1 LteAuthAmf:gAA= LteAuthOp:EREREREREREREREREREREQ== Mcc:001 Mnc:01 Mobility:[] NetworkServices:[policy_enforcement] RelayEnabled:0xc0000265dc SubProfiles:[] Tac:1}
I0911 22:37:40.081990     338 main.go:134] Old config ran: [123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]
I0911 22:37:40.082527     338 main.go:132] Old config: {Epc:0xc000372180 FegNetworkID:[] Ran:[123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]}
I0911 22:37:40.082654     338 main.go:133] Old config epc: &{CloudSubscriberdbEnabled:false DefaultRuleID:default_rule_1 LteAuthAmf:gAA= LteAuthOp:EREREREREREREREREREREQ== Mcc:001 Mnc:01 Mobility:[] NetworkServices:[policy_enforcement] RelayEnabled:0xc0000267a6 SubProfiles:[] Tac:1}
I0911 22:37:40.082891     338 main.go:134] Old config ran: [123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]
I0911 22:37:40.083287     338 main.go:164] New config: {Epc:0xc00034cb60 FegNetworkID:[] Ran:[123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]}
I0911 22:37:40.083442     338 main.go:165] New config epc: &{CloudSubscriberdbEnabled:false DefaultRuleID:default_rule_1 GxGyRelayEnabled:0xc0000265dc HssRelayEnabled:0xc0000265dc LteAuthAmf:gAA= LteAuthOp:EREREREREREREREREREREQ== Mcc:001 Mnc:01 Mobility:[] NetworkServices:[policy_enforcement] SubProfiles:[] Tac:1}
I0911 22:37:40.083649     338 main.go:166] New config ran: [123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]
I0911 22:37:40.084104     338 main.go:164] New config: {Epc:0xc00034d1e0 FegNetworkID:[] Ran:[123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]}
I0911 22:37:40.084266     338 main.go:165] New config epc: &{CloudSubscriberdbEnabled:false DefaultRuleID:default_rule_1 GxGyRelayEnabled:0xc0000267a6 HssRelayEnabled:0xc0000267a6 LteAuthAmf:gAA= LteAuthOp:EREREREREREREREREREREQ== Mcc:001 Mnc:01 Mobility:[] NetworkServices:[policy_enforcement] SubProfiles:[] Tac:1}
I0911 22:37:40.084368     338 main.go:166] New config ran: [123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125]
I0911 22:37:40.084471     338 main.go:181] [RUN] UPDATE cfg_network_configs SET value = $1 WHERE network_id = $2 AND type = $3 [[123 34 101 112 99 34 58 123 34 100 101 102 97 117 108 116 95 114 117 108 101 95 105 100 34 58 34 100 101 102 97 117 108 116 95 114 117 108 101 95 49 34 44 34 103 120 95 103 121 95 114 101 108 97 121 95 101 110 97 98 108 101 100 34 58 102 97 108 115 101 44 34 104 115 115 95 114 101 108 97 121 95 101 110 97 98 108 101 100 34 58 102 97 108 115 101 44 34 108 116 101 95 97 117 116 104 95 97 109 102 34 58 34 103 65 65 61 34 44 34 108 116 101 95 97 117 116 104 95 111 112 34 58 34 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 81 61 61 34 44 34 109 99 99 34 58 34 48 48 49 34 44 34 109 110 99 34 58 34 48 49 34 44 34 110 101 116 119 111 114 107 95 115 101 114 118 105 99 101 115 34 58 91 34 112 111 108 105 99 121 95 101 110 102 111 114 99 101 109 101 110 116 34 93 44 34 116 97 99 34 58 49 125 44 34 114 97 110 34 58 123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125 125] lte_network cellular_network]
I0911 22:37:40.085527     338 main.go:181] [RUN] UPDATE cfg_network_configs SET value = $1 WHERE network_id = $2 AND type = $3 [[123 34 101 112 99 34 58 123 34 100 101 102 97 117 108 116 95 114 117 108 101 95 105 100 34 58 34 100 101 102 97 117 108 116 95 114 117 108 101 95 49 34 44 34 103 120 95 103 121 95 114 101 108 97 121 95 101 110 97 98 108 101 100 34 58 102 97 108 115 101 44 34 104 115 115 95 114 101 108 97 121 95 101 110 97 98 108 101 100 34 58 102 97 108 115 101 44 34 108 116 101 95 97 117 116 104 95 97 109 102 34 58 34 103 65 65 61 34 44 34 108 116 101 95 97 117 116 104 95 111 112 34 58 34 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 82 69 81 61 61 34 44 34 109 99 99 34 58 34 48 48 49 34 44 34 109 110 99 34 58 34 48 49 34 44 34 110 101 116 119 111 114 107 95 115 101 114 118 105 99 101 115 34 58 91 34 112 111 108 105 99 121 95 101 110 102 111 114 99 101 109 101 110 116 34 93 44 34 116 97 99 34 58 49 125 44 34 114 97 110 34 58 123 34 98 97 110 100 119 105 100 116 104 95 109 104 122 34 58 50 48 44 34 116 100 100 95 99 111 110 102 105 103 34 58 123 34 101 97 114 102 99 110 100 108 34 58 52 52 53 57 48 44 34 115 112 101 99 105 97 108 95 115 117 98 102 114 97 109 101 95 112 97 116 116 101 114 110 34 58 55 44 34 115 117 98 102 114 97 109 101 95 97 115 115 105 103 110 109 101 110 116 34 58 50 125 125 125] nms_network cellular_network]
I0911 22:37:40.089148     338 main.go:85] SUCCESS
I0911 22:37:40.089191     338 main.go:86] END MIGRATION
```

Local entries in table `cfg_network_configs` before running migration:
```
pgdump-dev-9-4-20=# select network_id, type, encode(value, 'escape') from cfg_network_configs;
 network_id  |       type       |                                                                                                                                                           encode
-------------+------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 lte_network | cellular_network | {"epc":{"default_rule_id":"default_rule_1","lte_auth_amf":"gAA=","lte_auth_op":"EREREREREREREREREREREQ==","mcc":"001","mnc":"01","network_services":["policy_enforcement"],"relay_enabled":false,"tac":1},"ran":{"bandwidth_mhz":20,"tdd_config":{"earfcndl":44590,"special_subframe_pattern":7,"subframe_assignment":2}}}
 lte_network | dnsd_network     | {"enable_caching":false,"local_ttl":0}
 lte_network | orc8r_features   | {"features":{"placeholder":"true"}}
 nms_network | cellular_network | {"epc":{"default_rule_id":"default_rule_1","lte_auth_amf":"gAA=","lte_auth_op":"EREREREREREREREREREREQ==","mcc":"001","mnc":"01","network_services":["policy_enforcement"],"relay_enabled":false,"tac":1},"ran":{"bandwidth_mhz":20,"tdd_config":{"earfcndl":44590,"special_subframe_pattern":7,"subframe_assignment":2}}}
 nms_network | dnsd_network     | {"enable_caching":false,"local_ttl":0}
 nms_network | orc8r_features   | {"features":{"placeholder":"true"}}
(6 rows)
```

Local entries in table `cfg_network_configs` after running migration:
```
pgdump-dev-9-4-20=# select network_id, type, encode(value, 'escape') from cfg_network_configs;
 network_id  |       type       |                                                                                                                                                                           encode
-------------+------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 lte_network | dnsd_network     | {"enable_caching":false,"local_ttl":0}
 lte_network | orc8r_features   | {"features":{"placeholder":"true"}}
 nms_network | dnsd_network     | {"enable_caching":false,"local_ttl":0}
 nms_network | orc8r_features   | {"features":{"placeholder":"true"}}
 lte_network | cellular_network | {"epc":{"default_rule_id":"default_rule_1","gx_gy_relay_enabled":false,"hss_relay_enabled":false,"lte_auth_amf":"gAA=","lte_auth_op":"EREREREREREREREREREREQ==","mcc":"001","mnc":"01","network_services":["policy_enforcement"],"tac":1},"ran":{"bandwidth_mhz":20,"tdd_config":{"earfcndl":44590,"special_subframe_pattern":7,"subframe_assignment":2}}}
 nms_network | cellular_network | {"epc":{"default_rule_id":"default_rule_1","gx_gy_relay_enabled":false,"hss_relay_enabled":false,"lte_auth_amf":"gAA=","lte_auth_op":"EREREREREREREREREREREQ==","mcc":"001","mnc":"01","network_services":["policy_enforcement"],"tac":1},"ran":{"bandwidth_mhz":20,"tdd_config":{"earfcndl":44590,"special_subframe_pattern":7,"subframe_assignment":2}}}
(6 rows)
```

### Verifying orc8r APIs

GET /lte/lte_network
```
{"cellular":{"epc":{"default_rule_id":"default_rule_1","gx_gy_relay_enabled":false,"hss_relay_enabled":false,"lte_auth_amf":"gAA=","lte_auth_op":"EREREREREREREREREREREQ==","mcc":"001","mnc":"01","network_services":["policy_enforcement"],"tac":1},"ran":{"bandwidth_mhz":20,"tdd_config":{"earfcndl":44590,"special_subframe_pattern":7,"subframe_assignment":2}}},"description":"Lte Network Description","dns":{"enable_caching":false,"local_ttl":0},"features":{"features":{"placeholder":"true"}},"id":"lte_network","name":"Lte Network"}
```

## Additional Information

- [x] This change is backwards-breaking

### What breaks
A swagger model for `network_epc_configs` has been modified, and requires db migration to be run, otherwise cloud-side code will error out.

### REQUIRED FOR ONCALL
After staging orc8r is deployed, and after production orc8r is deployed, migration script `m013_relay_split` should be run.
To do this, ssh into any of the controller pods, navigate to `/var/opt/magma/bin` and run `./m013_relay_split`. You should see the following output in stdout:
```
SUCCESS
END MIGRATION
```

### CODE REVIEW

**Review by ssanadhya — APPROVED** [review]
Changes to MME (which are all from PR 2575) look good to me

**Review by themarwhal — APPROVED** [review]

**Comment by themarwhal** [comment]
looks like orc8r test if failing though

File: lte/cloud/go/tools/migrations/m013_relay_split/types/new.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	// epc
+	// Required: true
+	Epc *NetworkEpcConfigs `json:"epc"`
+
+	// feg network id
+	FegNetworkID FegNetworkID `json:"feg_network_id,omitempty"`
+
+	// ran
+	// Required: true
+	Ran *NetworkRanConfigs `json:"ran"`
```
Comment (xjtian) [inline lte/cloud/go/tools/migrations/m013_relay_split/types/new.go]: For the fields that you're not migrating, you should just use `json.RawMessage`

File: lte/cloud/go/tools/migrations/m013_relay_split/main.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	sqlStr, args, _ := b.ToSql()
+	glog.Infof("[RUN] %s %v", sqlStr, args)
+	rows, err := b.RunWith(tx).Query()
+	if err != nil {
+		return errors.Wrap(err, "get lte_network configs")
+	}
+	oldByNid := map[string]types.OldNetworkCellularConfigs{}
+	for rows.Next() {
+		var nid string
+		old := types.OldNetworkCellularConfigs{}
```
Comment (xjtian) [inline lte/cloud/go/tools/migrations/m013_relay_split/main.go]: declare this closer to where it's used

File: lte/cloud/go/tools/migrations/m013_relay_split/main.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+		newBytes, err := json.Marshal(newConf)
+		if err != nil {
+			return errors.Wrap(err, "marshal updated lte_network config")
+		}
+		updateConfByNid[nid] = newBytes
+	}
+
+	// Update lte_network configs
+	for nid := range updateConfByNid {
+		updateConf := updateConfByNid[nid]
```
Comment (xjtian) [inline lte/cloud/go/tools/migrations/m013_relay_split/main.go]: use the 2-arg version of range

File: orc8r/cloud/go/tools/migrations/m013_relay_split/main.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/*
```
Comment (xjtian) [inline orc8r/cloud/go/tools/migrations/m013_relay_split/main.go]: delete this file

**Review by xjtian — DISMISSED** [review]

File: orc8r/cloud/go/service/middleware/unary/identity_decorator.go [hunk 5]
```diff
@@ -277,8 +277,8 @@ func ensureLocalPeer(ctx context.Context) error {
 	if ip == nil {
 		return status.Errorf(codes.PermissionDenied, "Invalid Client Address: %+v", caller.Addr)
 	}
-	if !ip.IsLoopback() {
-		return status.Errorf(codes.PermissionDenied, "Missing Client Certificate from Client %s", ip.String())
-	}
+	//if !ip.IsLoopback() {
+	//	return status.Errorf(codes.PermissionDenied, "Missing Client Certificate from Client %s", ip.String())
+	//}
```
Comment (xjtian) [inline orc8r/cloud/go/service/middleware/unary/identity_decorator.go]: revert this

**Review by xjtian — DISMISSED** [review]

**Review by xjtian — APPROVED** [review]

**Pull request merged**